balancer/base/balancer.go

```diff
@@ -101,17 +101,40 @@ func (b *baseBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
 	// addrsSet is the set converted from addrs, it's used for quick lookup of an address.
 	addrsSet := make(map[resolver.Address]struct{})
 	for _, a := range s.ResolverState.Addresses {
-		addrsSet[a] = struct{}{}
-		if _, ok := b.subConns[a]; !ok {
+		// Strip attributes from addresses before using them as map keys. So
+		// that when two addresses only differ in attributes pointers (but with
+		// the same attribute content), they are considered the same address.
+		//
+		// Note that this doesn't handle the case where the attribute content is
+		// different. So if users want to set different attributes to create
+		// duplicate connections to the same backend, it doesn't work. This is
+		// fine for now, because duplicate is done by setting Metadata today.
+		//
+		// TODO: read attributes to handle duplicate connections.
+		aNoAttrs := a
+		aNoAttrs.Attributes = nil
+		addrsSet[aNoAttrs] = struct{}{}
+		if sc, ok := b.subConns[aNoAttrs]; !ok {
 			// a is a new address (not existing in b.subConns).
+			//
+			// When creating SubConn, the original address with attributes is
+			// passed through. So that connection configurations in attributes
+			// (like creds) will be used.
 			sc, err := b.cc.NewSubConn([]resolver.Address{a}, balancer.NewSubConnOptions{HealthCheckEnabled: b.config.HealthCheck})
 			if err != nil {
 				logger.Warningf("base.baseBalancer: failed to create new SubConn: %v", err)
 				continue
 			}
-			b.subConns[a] = sc
+			b.subConns[aNoAttrs] = sc
 			b.scStates[sc] = connectivity.Idle
 			sc.Connect()
+		} else {
+			// Always update the subconn's address in case the attributes
+			// changed. This is a noop if the current address is the same as the
+			// old one (reflect.DeepEqual).
+			//
+			// TODO: delete this when this balancer reads attributes.
```

Comment:
This is what Java does now.  It's unclear if we should be trying to be more aggressive about forcing a reconnect to pick up the new attributes if we know they changed.

Reply:
When we know how to hash attributes, and know which part of it matters for connections, it will be easier to decide a `UpdateAddresses` is needed. And the SubConn would decide if it will reconnect to pick up the new attributes.

Reply:
> will be easier to decide a UpdateAddresses is needed 

Right -- my point is we will still need to call `UpdateAddresses` in that case, so we won't be deleting this.

Reply:
I would still say we should delete the TODO.  We will update the addresses when the attributes change in a meaningful way.

Reply:
Done